In [1]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.naive_bayes import GaussianNB
import pandas as pd

df = pd.read_csv("iris.csv")

[train_set, test_set] = train_test_split(df.values, train_size=0.7, random_state=278839)

test_class = test_set[:, 4]
test_input = test_set[:, 0:4]
train_class = train_set[:, 4]
train_input = train_set[:, 0:4]

Przygotowanie danych

In [2]:
clf = KNeighborsClassifier(n_neighbors=3)
clf = clf.fit(train_input, train_class)

print(clf.score(test_input, test_class)*100)

y_true = test_class
y_pred = clf.predict(test_input)

print(confusion_matrix(y_true, y_pred))

95.55555555555556
[[16  0  0]
 [ 0 15  1]
 [ 0  1 12]]


K-najbliższych sąsiadów: 3

In [3]:
clf = KNeighborsClassifier(n_neighbors=5)
clf = clf.fit(train_input, train_class)

print(clf.score(test_input, test_class)*100)

y_true = test_class
y_pred = clf.predict(test_input)

print(confusion_matrix(y_true, y_pred))

95.55555555555556
[[16  0  0]
 [ 0 15  1]
 [ 0  1 12]]


K-najbliższych sąsiadów: 5

In [4]:
clf = KNeighborsClassifier(n_neighbors=11)
clf = clf.fit(train_input, train_class)

print(clf.score(test_input, test_class)*100)

y_true = test_class
y_pred = clf.predict(test_input)

print(confusion_matrix(y_true, y_pred))

95.55555555555556
[[16  0  0]
 [ 0 14  2]
 [ 0  0 13]]


K-najbliższych sąsiadów: 11

In [5]:
gnb = GaussianNB()

y_pred = gnb.fit(train_input, train_class).predict(test_input)

print("Number of mislabeled points out of a total %d points : %d" % (test_set.shape[0], (test_class != y_pred).sum()))
print(42/45*100)
print(confusion_matrix(y_true, y_pred))

Number of mislabeled points out of a total 45 points : 3
93.33333333333333
[[16  0  0]
 [ 0 14  2]
 [ 0  1 12]]


Naiwny bayes

## Podsumowanie
własne drzewko: 93.(3)%, 42 - dobrych

automatyczne drzewko: 93.(3)%, 42 - dobrych

k-3: 95.55555555555556%, 43 - dobrych

k-5: 95.55555555555556%, 43 - dobrych

k-3 i k-5 mylą się w tych samych miejscach

k-11: 95.55555555555556%, 43 - dobrych

k-11 myli się już w innych

gnb: 93.(3)%, 42 - dobrych

K-najbliższych sąsiadów, wszystkie dają takie same wyniki lepsze od własnego czy automatycznie wygenerowanego drzewka.
Naiwny bayes także daje takie same wyniki co własne i automatycznie wygenerowane drzewka ale gorsze od k-najbliższych sąsiadów